### PR TITLE
Implemented experimental::Sequencer

### PR DESCRIPTION
**Pull request by demin80**
Signed-off-by: Denis Mindolin <dmindolin1@bloomberg.net>

**Describe your changes**
This PR introduces a lightweight Sequencer module (class quantum::experimental::Sequencer) whose functionality is similar to quantum::Sequencer but has the following differences w.r.t. quantum::Sequencer.
* Ordering of pending tasks in quantum::Sequencer is done via quantum::Dispatcher. In particular, each tasks scheduled via quantum::Sequencer explicitly waits for its dependent tasks to finish. However while in the pending state, these tasks are still enqueued into quantum::Dispatcher. In the presence of a large number of dependent tasks, quantum::Dispatcher needs to keep traversing over those queues to find a tasks with no dependencies (i.e. ready to be executed). This often results in a scheduling lag (for the queue traversal) + wasted CPU cycles for checking the states of dependents. In contrast to that, quantum::experimental::Sequencer constructs a dependency graph outside of quantum::Dispatcher and schedules them via quantum::Dispatcher only when they have no pending dependents. 

**Testing performed**
Unit tests added

**Additional context**
Add any other context about your contribution here.

### CODE REVIEW

**Comment by arosenzweig3** [comment]
Just some preliminary comments for now as I haven't gotten to look deeply into the code:

1. We should try to support the flexible perfect forwarding of arguments approach that the `Dispatcher` and the original `Sequencer` use. This can be achieved directly via some lambda capture into tuple and std::apply. We also have some useful related functions in `quantum_capture_util.h` and `quantum_stl_impl.h`. While we could use better in-header examples of how to use this, it's the standard paradigm within the library and is better than requiring the user to coerce their function into the int(VoidContextPtr) via bind or lambdas. We discussed this offline as well.
2. We should add some additional documentation on `Sequencer` indicating the performance issues under high lead and referring to `SequencerLite` as an alternative. The description here in the PR is good, but we should replicate it in `quantum_sequencer.h` so that developers will be aware.
3. @albertdang2 raised a question when the class was first mentioned of what's missing from `SequencerLite` to make it Lite. We may want to brainstorm some other names.

**Comment by accelerated** [comment]
Will also try to take a look ...

**Comment by demin80** [comment]
@accelerated Alex, that would be great!
@arosenzweig3 Regarding your comment 3, the Lite suffix of the class name is to signify the fact that it's higher performant.

File: quantum/util/quantum_sequencer_lite.h
Comment (arosenzweig3) [inline quantum/util/quantum_sequencer_lite.h:52]: Let's document the exception callback here; it's existence is buried within the configuration and barely documented there. Just a brief on the fact that the sequencer will swallow exceptions and call that callback if specified should be good. Might be good to copy that note to the old sequencer as well.

File: quantum/util/quantum_sequencer_lite.h
Comment (arosenzweig3) [inline quantum/util/quantum_sequencer_lite.h:289]: For consistency I'd suggest we use the `quantum::Mutex` here and its associated guards in the implementation.

File: tests/quantum_sequencer_lite_tests.cpp
Comment (arosenzweig3) [inline tests/quantum_sequencer_lite_tests.cpp:63]: The asserts above should make this if superfluous; might as well remove it.

**Review by arosenzweig3 — COMMENTED** [review]
While I don't think it makes sense to build comparative benchmarks into the unit tests, it would be good to see the methods used to demonstrate that `SequencerLite` is less load-intensive than `Sequencer`.

**Review by accelerated — CHANGES_REQUESTED** [review]
This is a very nice Sequencer improvement and quite cleverly designed. I would have a class design recommendation (aside from the Configuration classes which could also be shared - as mentioned in an earlier comment along with the Json schemas). 

Since both `Sequencer` and `SequencerLite` share the exact same interface (as they should!) I would create a PIMPL design using `Sequencer`. The Impl inside would either be the `SequencerHeavy` (original) and another one being `SequencerLite`. Other sequencer implementations can be added later...The `Sequencer` constructor would use the Configuration and perhaps an enum indicating which instance to create, then create the internal impl. You can also add a protected method to inject the impl class directly so that you can test with either and just swap out the implementation at will.

Since you can now swap implementations, you should be able to execute the same test suite twice, no need for separate tests for `SequencerLite`. Also it would be nice to get some real benchmark performance differences between the two.

File: quantum/util/impl/quantum_sequencer_lite_impl.h
Comment (demin80) [reply quantum/util/impl/quantum_sequencer_lite_impl.h:130]: Let's use your example: we have two pending universal tasks T1 and T2 all enqueued into N queues + the universal queue. Then assume we start executing task T1, and once it's done, SequencerLite will call removePending for T1. Before the for-loop from line 128 starts, the T2's _pendingKeyCount will be equal to N (because T2 is not at the head of any of the N queues yet). Each call of removingPending for T1 will result in T2 becoming at the head of the corresponding queue and reducing T2->_pendingKeyCount  by one. However, those removePending will be always returning nullptr (see lines 114-116) until the final call that will make T2->_pendingKeyCount equal to 0 and hence returning it as the next task to execute. Hence T2 will be executed only once.

Comment (accelerated) [reply quantum/util/impl/quantum_sequencer_lite_impl.h:130]: Ah ok...good. I was not sure if this scenario was handled or not.

File: quantum/util/impl/quantum_sequencer_lite_impl.h
Comment (accelerated) [reply quantum/util/impl/quantum_sequencer_lite_impl.h:203]: There's a reason for using these utils, if you capture a non-copyable object in a lambda and you want to assign it to a `std::function`, `std::function` requires all captures to also be copyable. This is unfortunate so in order to allow movable-only types, we are using this instead.

**Comment by accelerated** [comment]
Also change the PR comment since now you can enqueue to any queue. The std::function restriction also does not apply in the sense that both Sequencer and SequencerLite interfaces are the same. If you go with the PIMPL design, maybe we can call the impl classes ChainedSequencer (orig) vs LazySequencer (lite) or ProactiveSequencer (orig - tasks are enqueued immediately) vs ReactiveSequencer (tasks are only enqueued when previous one completes)? I think SequencerLite does not really describe what it's doing. Anyway, just some ideas...

**Comment by arosenzweig3** [comment]
In general I think I'm broadly in favor of merging the configurations. I believe there's one config item unique to the current `Sequencer` that `SequencerLite` doesn't use. A shared configuration would also provide a reasonable way for users to decide whether to use the current implementation or the new one, now that the usage is essentially the same.

That said, I'm not sure there's much of a use case for the current implementation of `Sequencer`, beyond offering it to allow clients to see no change. I just ran `SequencerLite` through the same test program which led to pushing for its existence, in comparison to the current `Sequencer`. I did not include any memory footprint there, but on both CPU usage and actual time to complete the enqueued jobs, it massively outperformed `Sequencer`.

This test program has a simple job which takes 400-500 microseconds of churn (generating random numbers), and some instrumentation. In both cases, I set the program to enqueue 2,000 jobs per second for 300 seconds. With `Sequencer`, the CPU usage almost instantly climbed until it was using all available cores (in this case 5, as there were 5 coroutine threads), and with that it was only able to process roughly 10 jobs per second, due to the time spent surveying the coroutines that `Sequencer` had created. I killed the process rather than wait for all jobs to be completed. With `SequencerLite`, CPU usage went a little bit over one core, and finished the last job about 16 seconds after it had been enqueued.

I'll re-run this after @demin80 has finished adjusting to the prior comments, but it is extremely promising.

**Comment by arosenzweig3** [comment]
A secondary comparison that was available from my instrumentation which I did not include above: the median time between one job finishing and the next starting. This was actually marginally better for `SequencerLite` than for `Sequencer` under light load. Under heavy load, that metric for `Sequencer` was very poor.

**Comment by arosenzweig3** [comment]
One thing to note for reference: I have not run any of these tests with multiple sequence keys; they were all simple one-key tests. They did not cover `enqueueAll()`, `enqueue()` with a `vector` of keys, or even using multiple keys across the various `enqueue()` calls. They were very specifically about the simplest case. Similar analysis with multiple keys, etc. should also be done.

**Comment by arosenzweig3** [comment]
A side note, tied into @accelerated 's comments above; we found that the `wrap()` function `SequencerLite` is using is less capable than the general utilities `quantum` is already using; it should be replaced with one of those (or a new one of those); it cannot take all signatures, though it can take some. (I prototyped `wrap()` for @demin80 originally.)

File: tests/quantum_sequencer_lite_tests.cpp
Comment (smcinerney1) [reply tests/quantum_sequencer_lite_tests.cpp:63]: The asserts will likely not be present in production code.

File: quantum/util/quantum_sequencer_lite.h
Comment (demin80) [reply quantum/util/quantum_sequencer_lite.h:289]: using quantum::mutex requires to have a pointer to quantum::context. We don't have an active context in SeqeunecerLite::enqueue.

File: quantum/util/impl/quantum_sequencer_lite_impl.h
Comment (demin80) [reply quantum/util/impl/quantum_sequencer_lite_impl.h:124]: yes I was thinking how we can get rid of this std::mutex. I can't switch to quantum::Mutex yet because there's no context in SequencerLite::enqueue, and I need it to lock the mutex there for inserting the task to the queue. Notice that removePending is not very heavy and is actually a constant-time for non-universal tasks (provided the number of keys assigned to each task is a constant) . So locking the std::mutex here may not be a big deal. 

One way of switching to quantum::Mutex instead would be to enqueue each task to the control queue first and then in the control coroutine update the actual queues under the quantum::Mutex.

File: quantum/util/quantum_sequencer_lite.h
Comment (arosenzweig3) [reply quantum/util/quantum_sequencer_lite.h:289]: We should be able to use `quantum::local::context()` here; if I'm recalling correctly that will work correctly in a coroutine or out of one.

Comment (demin80) [reply quantum/util/quantum_sequencer_lite.h:289]: quantum::local::context() will return nullptr outside of a coroutine. quantum::local is for accessing coroutine local storage (similar to TLS (thread local storage) for threads).

Comment (arosenzweig3) [reply quantum/util/quantum_sequencer_lite.h:289]: Right, and if we try to lock the `quantum::Mutex` from outside a coroutine, calling `lock(nullptr)` is the same as calling `lock()`; that's what we want, and works regardless of whether we are in a coroutine or not.

File: quantum/util/impl/quantum_sequencer_lite_configuration_impl.h
Comment (demin80) [reply quantum/util/impl/quantum_sequencer_lite_configuration_impl.h:66]: std::queue provides no way of iterating through it. I need to traverse it in SequencerLite::addPendingTask

File: quantum/util/impl/quantum_sequencer_lite_impl.h
Comment (demin80) [reply quantum/util/impl/quantum_sequencer_lite_impl.h:73]: I still need to increment the pendingKeyCount counter...

**Review by arosenzweig3 — COMMENTED** [review]
Looks like we need a little tidying up on the configuration, but otherwise I believe I'm happy. Test coverage looks good. Tomorrow I'll grab this copy and run it through the same test program as before to re-verify the performance (which I have no doubts on) and verify the full versatility of the captures.

**Comment by arosenzweig3** [comment]
I ran a similar test with the latest iterations and unsurprisingly the `SequencerLite` performance remains vastly superior under heavy load. I also was able to get more function signatures into `enqueue()`.

Let's start looking at naming; I agree with @albertdang2 in offline discussion that `SequencerLite` is not a great name because it implies a light-weight version lacking some features. Perhaps if we're going to keep the old `Sequencer` around as well, we should rename `SequencerLite` based on the implementation; something referring to the directed acyclic graph it uses?

**Comment by accelerated** [comment]
I think keeping the old sequencer around would not make much sense if the performance is so much better in the new version and since the functionalities are identical. More so, some ppl might be tempted to use it and risk degrading their applications. Maybe it's best to just rename it as Sequencer and deprecate the previous one?

**Comment by arosenzweig3** [comment]
> I think keeping the old sequencer around would not make much sense if the performance is so much better in the new version and since the functionalities are identical. More so, some ppl might be tempted to use it and risk degrading their applications. Maybe it's best to just rename it as Sequencer and deprecate the previous one?

I'm inclined this way as well. I definitely don't want to have `SequencerLite` as the name because it implies 'lite' on features, not on performance problems. At this point I can't see anything that the original `Sequencer` offers aside from continuity. We could simply replace the implementation and mark the configuration controlQueueId functions as deprecated.

**Comment by arosenzweig3** [comment]
Summarizing the content of an offline discussion between myself, @albertdang2 , @safaruqi , and a couple others who I am not finding github accounts for:

1. The only real value we're seeing in the current `Sequencer` is continuity; we would like to be able to transition our applications using `Sequencer` from the current implementation to what is labeled `SequencerLite` here via application changes rather than exclusively via relinking. This gives us the ability to toggle (probably still with restarting the application, but without needing to change between builds against different versions of `quantum`).
2. Having a single `Sequencer` interface with the ability to choose the implementation is at least one way to accomplish this.
3. Longer-term, we can adjust the default implementation to be the new one, and/or remove the old/current implementation entirely.

I'm not convinced we'll have further sequencer implementations to look at, but given how complete a change this one is we'd like to be able to toggle the implementation by having both exist in `quantum` at least for a while.

**Comment by albertdang2** [comment]
We wanted application to be able to switch between old and new 'Sequencer' for initial deployment.  This means we would have `Sequencer` and `SquencerLite`.  I suggest we change name from `SequencerLite` to `FastSequencer`.   Once we see no need for two  `Sequencer`,  then the performant one will be the default `Sequencer`.

**Comment by demin80** [comment]
yes, now given that the functionalities of both sequencers are the pretty much the same, we could simply replace one with the other. However, we need to think about the rollout. We could also rename SequencerLite to SequencerExperimental and replace Seqeuencer with SequencerExperimental later.

**Comment by accelerated** [comment]
The sequencers can also be wrapped as `quantum::experimental::sequencer` (new one) and `quantum::sequencer` the old one. This follows more closely the STL paradigm with the experimental namespace.

**Comment by demin80** [comment]
> The sequencers can also be wrapped as `quantum::experimental::sequencer` (new one) and `quantum::sequencer` the old one. This follows more closely the STL paradigm with the experimental namespace.

Sounds good. Let's do this unless somebody objects. I am also going to eliminate SequencerConfigurationBase since there will be no need in it anymore.

**Comment by arosenzweig3** [comment]
> Sounds good. Let's do this unless somebody objects. I am also going to eliminate SequencerConfigurationBase since there will be no need in it anymore.

This sounds good to me. When we deprecate we can either delete the old `Sequencer`, or shove it into a `deprecated` namespace if we want it to remain in the code base.

**Review by safaruqi — COMMENTED** [review]
This is a very clever design of the sequencer. The use-case for coroutines is to wait for blocking calls, primarily for I/O - or to chain a bunch of steps of computation like we do in a state machine. In our case, although the sequencer sits before the write on Kafka (i.e. I/O), the ordering of the queued tasks itself is a CPU-bound operation that was being done via coroutines. The current solution takes care of that now as Denis has explained above. We can take this as a good lesson for ourselves in learning about using coroutines only where the problem presents itself for the utility they provide, and not to treat them like OS-level threads. They do have a huge scheduling overhead (per thread) in userspace.
I agree with keeping the old one as the main one still and putting the new one under quantum::experimental. We can then instantiate one or the other via a config. and would be able to switch to the old one in case some initial issues are found in the new one. A restart would be fine for this kind of rollback. There is no need to switch them dynamically at runtime. That would be too much work and would need significant changes in the surrounding code.

**Review by arosenzweig3 — COMMENTED** [review]
Overall I'm happy. A couple of comments in-line.

**Review by safaruqi — COMMENTED** [review]
Went through the code. The changes look good. This change to move the new sequencer to quantum::experimental namespace will make working with either implementation much easier now.

**Comment by arosenzweig3** [comment]
I'm happy here. I won't approve just yet, so that others have some time to review. I will approve if we get no more input after a few more days.

**Review by arosenzweig3 — APPROVED** [review]
Approving.

**Pull request merged**
